### PR TITLE
tikv_util: add timeout for `get_disk_space_stats`.

### LINKED ISSUE #17939: qps continued to drop to bottom during one of tikv io hang
### What version of TiKV are you using?
./tikv-server -V
 TiKV 
Release Version:   8.5.0
Edition:           Community
Git Commit Hash:   eb664852e123de68d450bc30e38419f688c5ee8b
Git Commit Branch: HEAD
UTC Build Time:    2024-11-28 09:40:17
Rust Version:      rustc 1.77.0-nightly (89e2160c4 2023-12-27)
Enable Features:   memory-engine pprof-fp jemalloc mem-profiling portable test-engine-kv-rocksdb test-engine-raft-raft-engine trace-async-tasks openssl-vendored
Profile:           dist_release
2024-12-04T18:43:02.191+0800	

### What operating system and CPU are you using?
- 6-tikv-32c/128g-m6g.8xlarge (io2-8000)

### Steps to reproduce
  1、aws op env
     - 3-tidb-32c/64g-c6g.8xlarge (gp3)
     - 6-tikv-32c/128g-m6g.8xlarge (io2-8000)
     - 1-pd-32c/64g-c6g.8xlarge (gp3)
     -  Cross-3 AZ
 2、run wetech walletCenter workload
  --db-driver=mysql --mysql-db=wetech_wallet --mysql-port=4000 --mysql-user=root --threads=128 --wallet_table_size=200000000 --wallet_table_partitions=800 --wallet_log_partition_size=5000000000 --wallet_log_partition_names=p1,p2,p3 --insert_update_ratio=0.1 --mysql-ignore-errors=all --report-interval=10 --time=0 wallet.lua run
3、injection one of tikv io hang lasts for 10mins

### What did you expect?
qps can recover within 5mins

### What did happened?
qps continued to drop to bottom during one of tikv io hang which lasts for 10mins until fault recover
Image

### PR DESCRIPTION
**Pull request by LykxSassinator**
### What is changed and how it works?

Issue Number: Close #17939

What's Changed:

Introduce a timeout for `get_disk_space_stats` to prevent the function from hanging indefinitely on I/O operations in case of I/O issues.

```commit-message
Introduce a timeout for `get_disk_space_stats` to prevent the function from hanging indefinitely on I/O operations in case of I/O issues.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
Introduce a timeout for `get_disk_space_stats` to prevent the function from hanging indefinitely on I/O operations in case of I/O issues.
```

### CODE REVIEW

File: components/tikv_util/src/sys/disk.rs [hunk 1]
```diff
@@ -82,15 +87,40 @@ pub fn get_disk_status(_store_id: u64) -> DiskUsage {
...
+    const DISK_STATS_CHECK_TIMEOUT: Duration = Duration::from_secs(5);
+
+    // Use a timeout channel to avoid blocking the thread when checking disk space.
+    let path = path.as_ref().to_owned();
+    let (tx, rx) = mpsc::channel();
+
+    // It's ok to use a separate thread to check disk space stats, because the
+    // access to disk is not frequent, and the disk check is not the hot path
+    // of tikv, so the overhead of allocating a thread is acceptable.
+    thread::spawn(move || {
```
Comment (Connor1996) [inline components/tikv_util/src/sys/disk.rs:105]: It would create and destroy threads too frequently, better to use a worker

**Comment by glorv** [comment]
What about `raft_engine.get_engine_size` and `kv_engine.get_engine_used_size` seems they may also involves IO operations:

https://github.com/tikv/tikv/blob/0659fcc574901dbce94b5ddf6d03d04b539f40f0/components/raftstore/src/store/worker/pd.rs#L2509-L2533

In general, I think we should either change to use async non-blocking IO interface(if possible) or decouple IO operations from pd worker to a separate thread.

File: components/tikv_util/src/sys/disk.rs (hunk 1 above)
Comment (LykxSassinator) [reply components/tikv_util/src/sys/disk.rs:105]: IMO, the frequency is low and acceptable, as this function is primarily called by:
1. StoreHeartbeat, which runs at a default interval of 10 seconds.
2. To check whether the disk is full or not, also at an interval of 10 seconds.

**Comment by LykxSassinator** [comment]
> decouple IO operations from pd worker to a separate thread.

How about reusing the `StoreStatsReporter::report_store_infos` ? We can make the individual thread `StatsMonitor` to collect the size of engines (kvdb and raft-engine) periodically.

**Comment by LykxSassinator** [comment]
Closed. Refactored by https://github.com/tikv/tikv/pull/17947.

**Pull request closed without merging**